### PR TITLE
ShipEngine address validation service method

### PR DESCRIPTION
**Pull request by pelargir**
This implements a service method for ShipEngine's address validation API endpoint.

### CODE REVIEW

File: lib/friendly_shipping/services/ship_engine/serialize_address_validation_request.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+              company_name: location.company_name,
+              address_line1: location.address1,
+              address_line2: location.address2,
+              address_line3: location.address3,
+              city_locality: location.city,
+              state_province: location.region.code,
+              postal_code: location.zip,
+              country_code: location.country.code,
+              address_residential_indicator: location.address_type == "residential" ? "yes" : "no"
+            }
```
Comment (davecandlescience) [inline lib/friendly_shipping/services/ship_engine/serialize_address_validation_request.rb:24]: It doesn't error out if any of the attributes are nil like UPS does?? :)

File: lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+              end
+              Success(
+                ApiResult.new(
+                  locations,
+                  original_request: request,
+                  original_response: response
+                )
+              )
+            end
+          end
```
Comment (davecandlescience) [inline lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb:54]: What do you think about adding `rescue JSON::ParserError` here? I can't remember how ShipEngine returns errors when they are down or having issues.

File: lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+module FriendlyShipping
+  module Services
+    class ShipEngine
+      class ParseAddressValidationResponse
+        extend Dry::Monads::Result::Mixin
+
+        class << self
+          # @param [FriendlyShipping::Request] request
+          # @param [FriendlyShipping::Response] response
+          # @return [Success<ApiResult>, Failure<ApiFailure>]
```
Comment (davecandlescience) [inline lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb:14]: Should this be Success<ApiResult<Array<Physical::Location>>>?

**Review by davecandlescience — APPROVED** [review]
Looks good! I do think it would be worth protecting against malformed responses from ShipEngine though.

File: lib/friendly_shipping/services/ship_engine/serialize_address_validation_request.rb (hunk 1 above)
Comment (pelargir) [reply lib/friendly_shipping/services/ship_engine/serialize_address_validation_request.rb:24]: Nope. Shocking eh?

File: lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb (hunk 2 above)
Comment (pelargir) [reply lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb]: That's a good idea.

File: lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb (hunk 3 above)
Comment (pelargir) [reply lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb]: Sure. More detail won't hurt.

**Pull request merged**